### PR TITLE
Handle overflows while calculating cost of the block in gas-metering algorithm

### PR DESCRIPTION
**Pull request by techraed**
Copy from https://github.com/paritytech/wasm-instrument/pull/29

### CODE REVIEW

**Review by ukint-vs — APPROVED** [review]

**Comment by breathx** [comment]
As we already discussed with @SabaunT,
I offer approach where `Cost { overflows: Vec<u8>, remainder: u32 }`, where sum of `overflows` elements equals amount of times we should call `gas(u32::MAX)`

It's more optimal by memory and reduces amount of allocations in overflow cases.

For example, if we ran into 5 overflows, current approach will twice reallocate vec and contain `[u32::MAX, u32::MAX, u32::MAX, u32::MAX, u32::MAX, reminder]`, while offered one: `Cost { overflows: [5], reminder }`

**Review by breathx — APPROVED** [review]

**Review by grishasobol — CHANGES_REQUESTED** [review]
Whether this solution is checked on https://github.com/gear-tech/gear/issues/1353

File: src/gas_metering/mod.rs [hunk 1]
```diff
@@ -480,6 +581,15 @@ fn add_grow_counter<R: Rules>(
 	b.build()
 }
 
+fn inject_counter<R: Rules>(
```
Comment (grishasobol) [inline src/gas_metering/mod.rs]: what the reason to move this function?

File: src/gas_metering/mod.rs [hunk 2]
```diff
@@ -284,7 +284,105 @@ struct MeteredBlock {
...
+///
+/// The latter optimization approach is implemented and 2 fields are introduced:
+/// `max_out_overflows_num` and `overflows`.
+#[derive(Debug, PartialEq, PartialOrd)]
+#[cfg_attr(test, derive(Copy, Clone, Default))]
+struct BlockCostCounter {
+	/// Amount of overflow counters that overflowed themselves.
+	/// For more info see type docs.
+	max_out_overflows_num: usize,
+	/// Current amount of overflows.
```
Comment (grishasobol) [inline src/gas_metering/mod.rs]: Why we cant just use `u64` here instead all of this fields. Then in gas injection we will split u64 into two parts [u32, u32] and then we can use it just like all this fields

**Review by grishasobol — CHANGES_REQUESTED** [review]

File: src/gas_metering/mod.rs (hunk 1 above)
Comment (techraed) [reply src/gas_metering/mod.rs]: For readability reason. See "Clean code" of Robert Martin. Or: https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29 (Source code structure, rule 2).

Comment (grishasobol) [reply src/gas_metering/mod.rs]: Ok, but it's not our code, so we must change nothing except what we really need to change.

**Comment by techraed** [comment]
> Whether this solution is checked on [gear-tech/gear#1353](https://github.com/gear-tech/gear/issues/1353)

Yes, it was tested with current changes and works good. By the way, there is a new `cost_overflow` test in `gas_metering` module, which tests the expected set of instructions in case of overflows.

File: src/gas_metering/mod.rs (hunk 2 above)
Comment (techraed) [reply src/gas_metering/mod.rs]: discussed

**Review by grishasobol — APPROVED** [review]

**Pull request merged**